### PR TITLE
Unstub validator rpc

### PR DESCRIPTION
**Pull request by mpetrunic**
I started unstubbing validator rpc calls.
I'm not really sure how much of that will end up in beacon/validator api in beacon node so I've put abstract class that implements all those data processing functions that validator require.

I would leave them for now like this until rpc endpoints are defined and then we can either move this logic from `AbstractClient` to beacon node. If they are going to stay in validator it should probably be extracted to some class with more sane name.

I would like if you could check if I implemented those things correctly before I start writing tests.

### CODE REVIEW

**Comment by GregTheGreek** [comment]
Why did you re-implement all the functions? If anything you could import them from the beacon chain since its in the same project.

Also - check out my pr #185 because the RPC methods you'll need are a lot lighter

**Comment by mpetrunic** [comment]
> Why did you re-implement all the functions? If anything you could import them from the beacon chain since its in the same project.
> 
> Also - check out my pr #185 because the RPC methods you'll need are a lot lighter

TBH I failed to find some of this methods in chain folder, my bad (was late when I was searching)  :(

Do you plan exposing get_committee_assignment as validator rpc or should we calculate it in validator code? Depending on that I will either change to import from chain directory or make a call to API.

**Comment by GregTheGreek** [comment]
> > Why did you re-implement all the functions? If anything you could import them from the beacon chain since its in the same project.
> > Also - check out my pr #185 because the RPC methods you'll need are a lot lighter
> 
> TBH I failed to find some of this methods in chain folder, my bad (was late when I was searching) :(

Ahhh dont worry

> Do you plan exposing get_committee_assignment as validator rpc or should we calculate it in validator code? Depending on that I will either change to import from chain directory or make a call to API.

Well I think the validator wouldn't need that because the assingment would be passed in via the `get_duties` rpc call

**Comment by mpetrunic** [comment]
> Well I think the validator wouldn't need that because the assingment would be passed in via the `get_duties` rpc call

My chain of thought was to offload some of the computing from beacon node as all these stuff can be calculated from BeaconState.

**Comment by GregTheGreek** [comment]
> My chain of thought was to offload some of the computing from beacon node as all these stuff can be calculated from BeaconState.

Hmm.... I see what you're saying, I guess it matters on how "stateless" we want to be.

Also bare in mind that #185 is loosely implemented off of ethereum/eth2.0-specs#1011

File: src/validator/rpc/interface.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import {BeaconBlock, BeaconState, BLSPubkey, Epoch, Shard, Slot, ValidatorIndex} from "../../types";
+import {CommitteeAssignment, GenesisInfo} from "../types";
+
+export interface RpcClient {
```
Comment (wemeetagain) [inline src/validator/rpc/interface.ts:4]: I think this interface should not redefine the API of BeaconAPI, but rather inherit from it. If the validator needs more functionality, we should extend BeaconAPI (or ValidatorAPI or whatever we decide to call it).
The only thing we should need to extend/switch on here is how the RPC is communicated.

Comment (mpetrunic) [reply src/validator/rpc/interface.ts:4]: Agreed. I think validator now only needs ValidatorApi so Rpc client should only extend ValidatorApi interface.

**Comment by mpetrunic** [comment]
How about this interface for validator client?
I guess if we use produceBlock and produceAttestation, I could replace `onNewBeaconState` and `onNewAttestationdata` with `onNewValidatorDuty`?
@GregTheGreek @wemeetagain 
I will hold on to implementation till we close rpc PR. :)

**Comment by GregTheGreek** [comment]
Yeah lets take a back seat on this, I'm changing the api again

**Comment by GregTheGreek** [comment]
This is probably good to finish

**Pull request closed without merging**